cachi2/core/package_managers/bundler/parser.py

```diff
@@ -80,6 +80,44 @@ def download_to(self, deps_dir: RootedPath) -> None:
         download_binary_file(self.remote_location, fs_location)
 
 
+class GemPlatformSpecificDependency(GemDependency):
+    """
+    Represents a gem dependency built for a specific platform.
+
+    Attributes:
+        platform:     Platform for which the dependency was built.
+    """
+
+    platform: str
+
+    @property
+    def remote_location(self) -> str:
+        """Return remote location to download this gem from."""
+        return f"{self.source}/downloads/{self.name}-{self.version}-{self.platform}.gem"
+
+    def download_to(self, deps_dir: RootedPath) -> None:
+        """Download represented gem to specified file system location."""
+        fs_location = deps_dir.join_within_root(
+            Path(f"{self.name}-{self.version}-{self.platform}.gem")
+        )
+        log.info(
+            "Downloading platform-specific gem %s-%s-%s", self.name, self.version, self.platform
+        )
+        try:
+            download_binary_file(self.remote_location, fs_location)
+        except FetchError:
+            # A combination of Ruby v.3.0.7 and some Bundler dependencies results in
+            # -gnu suffix being dropped from some platforms. This was observed on
+            # sqlite3-aarch-linux-gnu. We cannot control user's Ruby version,
+            # but we could try and guess correct path. If this fails then we should
+            # assume that package is broken.
```

Comment:
What test data did you use? How did you come to the conclusion it has something to do with Ruby v3.0.7? When I tried the following `Gemfile` with that Ruby version (bundler 2.2.33):
```
source "https://rubygems.org"

gem "sqlite3", "= 2.0.4"
```
The lockfile I got for mult-arch platforms was (note the `mini_portile` dep):
```
GEM
  remote: https://rubygems.org/
  specs:
    mini_portile2 (2.8.7)
    sqlite3 (2.0.4)
      mini_portile2 (~> 2.8.0)

PLATFORMS
  aarch64-linux
  arm-linux
  arm-linux-musl
  arm64-darwin
  x64-mingw-ucrt
  x86-linux
  x86_64-darwin
  x86_64-linux
  x86_64-linux-musl

DEPENDENCIES
  sqlite3 (= 2.0.4)
```

I was not able to reproduce otherwise with newer Ruby which apparently changed how the compile targets are handled and with 3.1.0 I got:
```
GEM
  remote: https://rubygems.org/
  specs:
    sqlite3 (2.0.4-aarch64-linux-gnu)
    sqlite3 (2.0.4-arm-linux-gnu)
    sqlite3 (2.0.4-arm-linux-musl)
    sqlite3 (2.0.4-arm64-darwin)
    sqlite3 (2.0.4-x64-mingw-ucrt)
    sqlite3 (2.0.4-x86-linux-gnu)
    sqlite3 (2.0.4-x86_64-darwin)
    sqlite3 (2.0.4-x86_64-linux-gnu)
    sqlite3 (2.0.4-x86_64-linux-musl)

PLATFORMS
  aarch64-linux
  arm-linux
  arm-linux-musl
  arm64-darwin
  x64-mingw-ucrt
  x86-linux
  x86_64-darwin
  x86_64-linux
  x86_64-linux-musl

DEPENDENCIES
  sqlite3 (= 2.0.4)
```

That said, I found this issue which reported a similar problem, but for the `musl` target: https://github.com/rubygems/rubygems/issues/7432 . I wonder if these could be related in any way.

Reply:
The logic was eventually dropped -> resolving.